src/arkode/arkode_lsrkstep.c

```diff
@@ -0,0 +1,1436 @@
+  *dsmPtr   = ZERO;
+
+  /* access ARKodeLSRKStepMem structure */
+  retval = lsrkStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  sunrealtype rs  = (sunrealtype)step_mem->reqstages;
+  sunrealtype rn  = SUNRsqrt(rs);
+  sunrealtype rat = ONE / (rs - rn);
+  int in          = (int)rn;
+
+  /* Call the full RHS if needed. If this is the first step then we may need to
+     evaluate or copy the RHS values from an  earlier evaluation (e.g., to
+     compute h0). For subsequent steps treat this RHS evaluation as an
+     evaluation at the end of the just completed step to potentially reuse
+     (FSAL methods) RHS evaluations from the end of the last step. */
+
+  if (!(ark_mem->fn_is_current) && ark_mem->initsetup)
+  {
+    retval = step_mem->fe(ark_mem->tn, ark_mem->yn, ark_mem->fn,
+                          ark_mem->user_data);
+    step_mem->nfe++;
+    ark_mem->fn_is_current = SUNTRUE;
+    if (retval != ARK_SUCCESS) { return (ARK_RHSFUNC_FAIL); }
+  }
+
+  /* A tentative solution at t+h is returned in
+     y and its slope is evaluated in temp1.  */
+
+  N_VLinearSum(ONE, ark_mem->yn, ark_mem->h * rat, ark_mem->fn, ark_mem->ycur);
+  N_VLinearSum(ONE, ark_mem->yn, ark_mem->h / rs, ark_mem->fn, ark_mem->tempv1);
+
+  /* Evaluate stages j = 2,...,step_mem->reqstages */
+  for (int j = 2; j <= (int)((in - 1) * (in - 2) / 2); j++)
+  {
+    retval =
+      step_mem->fe(ark_mem->tcur + ((sunrealtype)j - ONE) * rat * ark_mem->h,
+                   ark_mem->ycur, ark_mem->fn, ark_mem->user_data);
+    if (retval != ARK_SUCCESS) { return (ARK_RHSFUNC_FAIL); }
+
+    N_VLinearSum(ONE, ark_mem->ycur, ark_mem->h * rat, ark_mem->fn,
+                 ark_mem->ycur);
+    N_VLinearSum(ONE, ark_mem->tempv1, ark_mem->h / rs, ark_mem->fn,
+                 ark_mem->tempv1);
+  }
+  N_VScale(ONE, ark_mem->ycur, ark_mem->tempv2);
+  for (int j = (int)((in - 1) * (in - 2) / 2 + 1);
+       j <= (int)(in * (in + 1) / 2 - 1); j++)
+  {
+    retval =
+      step_mem->fe(ark_mem->tcur + ((sunrealtype)j - ONE) * rat * ark_mem->h,
+                   ark_mem->ycur, ark_mem->fn, ark_mem->user_data);
+    if (retval != ARK_SUCCESS) { return (ARK_RHSFUNC_FAIL); }
+
+    N_VLinearSum(ONE, ark_mem->ycur, ark_mem->h * rat, ark_mem->fn,
+                 ark_mem->ycur);
+    N_VLinearSum(ONE, ark_mem->tempv1, ark_mem->h / rs, ark_mem->fn,
+                 ark_mem->tempv1);
+  }
+  retval = step_mem->fe(ark_mem->tcur +
+                          rat * (rn * (rn + ONE) / TWO - ONE) * ark_mem->h,
+                        ark_mem->ycur, ark_mem->fn, ark_mem->user_data);
+  if (retval != ARK_SUCCESS) { return (ARK_RHSFUNC_FAIL); }
+
+  N_VLinearSum(rn / (TWO * rn - ONE), ark_mem->tempv2,
+               (rn - ONE) / (TWO * rn - ONE), ark_mem->ycur, ark_mem->ycur);
+  N_VLinearSum(ONE, ark_mem->ycur,
+               (rn - ONE) * rat * ark_mem->h / (TWO * rn - ONE), ark_mem->fn,
+               ark_mem->ycur);
```

Comment:
Replace these two calls t o `N_VLinearSum` with a single call to `N_VLinearCombination`